googletest/include/gtest/internal/gtest-internal.h

```diff
@@ -1040,33 +1041,6 @@ Iter ArrayAwareFind(Iter begin, Iter end, const Element& elem) {
   return end;
 }
 
-// CopyArray() copies a k-dimensional native array using the elements'
-// operator=, where k can be any integer >= 0.  When k is 0,
-// CopyArray() degenerates into copying a single value.
-
-template <typename T, typename U>
-void CopyArray(const T* from, size_t size, U* to);
-
-// This generic version is used when k is 0.
-template <typename T, typename U>
-inline void CopyArray(const T& from, U* to) { *to = from; }
-
-// This overload is used when k >= 1.
-template <typename T, typename U, size_t N>
-inline void CopyArray(const T(&from)[N], U(*to)[N]) {
```

Comment:
This is why it worked. You begin with char[2][3] and through the call chain you are given char*[3]. So std::copy_n will dereference it, leaving it with char[3], hence the error of unassignability.

I like your idea with std::copy_n but that won't help recursion. If I'm not mistaken with C++17 it could just be:
```c++
template <typename T, typename U>
void CopyArray(const T* from, size_t size, U* to) {
    if constexpr (std::is_array<T>::value) // we cannot use if constexpr yet
        CopyArray(*from, std::extent<T>::value, *to);
    else
        std::copy_n(from, size, to);
}
```
I think you can change the existing code only a bit :<

Reply:
Thanks havent made time to look at it yet. Ill fix it soon